x/pairing/keeper/unresponsive_provider.go

```diff
@@ -111,84 +109,118 @@ func (k Keeper) getUnresponsiveProvidersToPunish(ctx sdk.Context, currentEpoch u
-		if err != nil {
-			return 0, utils.LavaError(ctx, k.Logger(ctx), "get_provider_payment_storage_from_stake_storage_list", map[string]string{"err": err.Error(), "epoch": fmt.Sprintf("%+v", previousEpoch)}, "couldn't get providerPaymentStorageList from stakeStorageList")
+		// try getting providerPaymentStorage using the providerPaymentStorageKey
+		providerPaymentStorage, found := k.GetProviderPaymentStorage(ctx, providerPaymentStorageKey)
+		if !found {
+			// Get previous epoch (from epochTemp)
+			previousEpoch, err := k.epochStorageKeeper.GetPreviousEpochStartForBlock(ctx, epochTemp)
+			if err != nil {
+				return utils.LavaError(ctx, k.Logger(ctx), "get_previous_epoch", map[string]string{"err": err.Error(), "epoch": fmt.Sprintf("%+v", epochTemp)}, "couldn't get previous epoch")
+			}
+			// update epochTemp
+			epochTemp = previousEpoch
+
+			// there is no providerPaymentStorage for this provider in this epoch -> continue searching
+			continue
 		}
 
-		// Go over providerPaymentStorage list
-		for _, providerPaymentStorage := range providerPaymentStorageList {
+		// counter is smaller than epochsNumToCheckCUForUnresponsiveProvider -> only count CU serviced by the provider in the epoch
+		if counter < epochsNumToCheckCUForUnresponsiveProvider {
+			// count the CU by iterating through the uniquePaymentStorageClientProvider objects
 			cuCount := uint64(0)
+			for _, uniquePayment := range providerPaymentStorage.GetUniquePaymentStorageClientProvider() {
+				cuCount += uniquePayment.GetUsedCU()
+			}
 
-			// Get provider address and chainID
-			providerAddress := k.getProviderAddressFromProviderPaymentStorageKey(providerPaymentStorage.Index)
-			chainID := k.getChainIDFromProviderPaymentStorageKey(providerPaymentStorage.Index)
-
-			// Create a punishEntryKey from the provider's address and the chain ID
-			punishEntryKey := providerAddress + "_" + chainID
-
-			// find the provider's punishEntry in map
-			providerPunishEntry, punishEntryFound := providersPunishEntryMap[punishEntryKey]
-
-			if isProvider {
-				// if isProvider is true, we count the CU serviced by the unersponsive provider
-				for _, uniquePayment := range providerPaymentStorage.GetUniquePaymentStorageClientProvider() {
-					cuCount += uniquePayment.GetUsedCU()
-				}
-				if punishEntryFound {
-					// punishEntry found -> add the cuCount to providerServicedCu and update the map entry
-					providerPunishEntry.providerServicedCu += cuCount
-					providersPunishEntryMap[punishEntryKey] = providerPunishEntry
-				} else {
-					// punishEntry not found -> create a new one and add it to the map
-					providersPunishEntryMap[punishEntryKey] = punishEntry{complainersUsedCu: 0, providerServicedCu: cuCount, providerPaymentStorageIndexList: []string{providerPaymentStorage.Index}}
-				}
+			// update the CU in the providerCuCounterForUnreponsivenessMap
+			if createNewproviderCuCounterForUnreponsivenessEntryFlag {
```

Comment:
remove this and create at the begining of the function

Reply:
same as previous comment